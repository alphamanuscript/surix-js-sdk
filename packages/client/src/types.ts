import { WrappedEntity, WrappedFile } from '@surix/data-helpers';

export interface Project {
  entities: ProjectEntities;
  files: ProjectFiles;
  tags: ProjectTags;
}

export interface ProjectEntities {
  get(id: string): Promise<WrappedEntity>;
  query(query?: Query): Promise<WrappedEntity[]>;
  create(entity: Entity): Promise<WrappedEntity>;
  put(entity: Entity): Promise<WrappedEntity>;
}

export interface ProjectFiles {
  get(id: string): Promise<WrappedFile>;
  list(): Promise<WrappedFile[]>;
}

export interface ProjectTags {
  list(): Promise<TagList>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];

export enum FieldType {
  TEXT = 'text',
  BOOLEAN = 'boolean',
  NUMBER = 'number',
  DATETIME = 'datetime',
  PHONE = 'phone',
  EMAIL = 'email',
  FILE = 'file',
  OBJECT = 'object',
  LIST = 'list'
}

export interface FieldPair {
  type: FieldType;
  value: RawEntityData | string | RawEntityData[] | number;
}

export interface RawEntityField extends FieldPair {
}

export interface RawEntityData {
  [fieldName: string]: RawEntityField;
}

/**
 * this interface describe the shape of the entity
 * as used within the system and the API
 */
export interface RawEntity {
  data: RawEntityData;
  tags?: string[];
}

export interface DataField {
  [fieldName: string]: string | number | object | boolean;
}
export interface Entity {
  data: DataField;
  tags?: string[];
  _id?: string;
}