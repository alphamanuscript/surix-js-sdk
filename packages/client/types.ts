import { EntityWrapper } from '@surix/data-helpers';

export interface Project {
  db: ProjectDatabase;
  files: ProjectFiles;
}

export interface ProjectDatabase {
  get(id: string): Promise<EntityWrapper>;
  query(query: Query): Promise<EntityWrapper[]>;
  getTags(): Promise<TagList[]>;
}

export interface ProjectFiles {
  get(id: string): Promise<ApiFile>;
  list(): Promise<ApiFile[]>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface ApiFile {
  name: string;
  status: 'failed' | 'ready' | 'pending';
  public: boolean;
  downloadUrl: string;
  tags: TagList[];
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];