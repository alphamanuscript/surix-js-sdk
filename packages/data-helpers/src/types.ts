export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldPathArray = Array<string|number>;

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | FileValue | ListValue | ObjectValue;

export interface FileValue {
    ref: string;
    mimeType: string;
    downloadUrl: string;
}

export type ListValue = ListItemValue[];

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: Field;
}

export interface Field extends TypeValuePair {
    name: string;
    label: string;
}

export interface TypeValuePair {
    type: FieldType;
    value: FieldValue;
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}