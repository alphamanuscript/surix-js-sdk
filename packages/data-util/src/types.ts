export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | ListItemValue[] | ObjectValue;

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: {
        name: string;
        type: FieldType;
        label: string;
        value: FieldValue;
    };
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}