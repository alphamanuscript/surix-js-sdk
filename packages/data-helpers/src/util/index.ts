export {
  deflateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkFieldPath,
  expandQuery
} from './util';