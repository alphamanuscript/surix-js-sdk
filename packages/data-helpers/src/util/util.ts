import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue, 
  TypeValuePair} from '../types';

type WalkFieldArg = TypeValuePair | Field;
type WalkFieldReturn = TypeValuePair | Field | undefined;

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath ? fieldPath.split('.') : [];
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...pathRemaining] = path;
  if (firstKey in entity.data) {
    return walkFieldPath(entity.data[firstKey], pathRemaining);
  }
}

export function walkFieldPath (field: WalkFieldArg, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return field;
  }
  if (field.type === 'object' || field.type === 'list') {
    const value = field.value as ObjectValue | ListValue;
    if (firstKey in value) {
      const subField = value[firstKey];
      return walkFieldPath(subField, restPath);
    }
  }
}

export function deflateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return deflateList(value);
    case 'object':
      return deflateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function deflateText (textValue: string): string {
  return textValue;
}

export function deflateNumber (numValue: string): number {
  return Number(numValue);
}

export function deflateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: deflateValue(value, type)
      };
    }, {});
}

export function deflateList (list: ListValue): any[] {
  return list.map(({ type, value }) => deflateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}