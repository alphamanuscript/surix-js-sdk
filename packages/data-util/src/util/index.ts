export {
  dehydrateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkFieldPath
} from './util';