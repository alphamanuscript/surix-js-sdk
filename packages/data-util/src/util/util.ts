import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue } from '../types';

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath.split('.');
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): any {
  return walkObjectPath(entity.data, path);
}

export function walkObjectPath (obj: ObjectValue | ListValue, path: FieldPathArray): any {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return Array.isArray(obj) ? dehydrateList(obj) : dehydrateObject(obj);
  }
  if (firstKey in obj) {
    const field = obj[firstKey];
    if (field.type === 'object' || field.type === 'list') {
      return walkObjectPath(field.value, restPath);
    }
    return dehydrateValue(field.value, field.type);
  }
}

export function dehydrateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return dehydrateList(value);
    case 'object':
      return dehydrateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function dehydrateText (textValue: string): string {
  return textValue;
}

export function dehydrateNumber (numValue: string): number {
  return Number(numValue);
}

export function dehydrateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: dehydrateValue(value, type)
      };
    }, {});
}

export function dehydrateList (list: ListValue): any[] {
  return list.map(({ type, value }) => dehydrateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}