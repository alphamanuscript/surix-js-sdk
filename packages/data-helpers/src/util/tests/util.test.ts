import * as util from '../util';

import { ApiEntity, Field, ObjectValue } from '../../types';

describe('util', () => {
  describe('makeField', () => {
    it('should create a field based on the give value and metadata', () => {
      const field = util.makeField('value', 'text');
      expect(field).toEqual({
        type: 'text',
        value: 'value'
      });
    });
  });

  describe('deflateValue', () => {
    describe('when value type is text', () => {
      it('should return the string text value', () => {
        expect(util.deflateValue('value', 'text')).toBe('value');
      });
    });
    describe('when value type is boolean', () => {
      it('should return the boolean value', () => {
        expect(util.deflateValue(true, 'boolean')).toBe(true);
      });
    });
    describe('when value type is number', () => {
      it('should the value cast to Number', () => {
        expect(util.deflateValue('120', 'number')).toBe(120);
      });
    });
    describe('when value type is datetime', () => {
      it('should the value cast to a Date', () => {
        const value = '2017-02-22T09:19:33.885Z';
        expect(util.deflateValue(value, 'datetime')).toEqual(new Date(value));
      });
    });
    describe('when value type is file', () => {
      it('should return the same file value', () => {
        const value = {
          ref: 'file21',
          mimeType: 'image/png',
          downloadUrl: '/download'
        };
        expect(util.deflateValue(value, 'file')).toEqual(value);
      });
    });
    describe('when value type is object', () => {
      it('should convert the value to plain object values without field metadata', () => {
        const value = {
          title: util.makeField('My Title', 'text'),
          read: util.makeField(true, 'boolean'),
          cover: util.makeField({
            ref: 'file1',
            mimeType: 'image/jpeg',
            downloadUrl: 'download1'
          }, 'file')
        };
        expect(util.deflateValue(value, 'object')).toEqual({
          title: 'My Title',
          read: true,
          cover: {
            ref: 'file1',
              mimeType: 'image/jpeg',
              downloadUrl: 'download1'
          }
        });
      });
    });
    describe('when value type is list', () => {
      it('should convert value to plain array of values without field metadata', () => {
        const value = [
          { type: 'text', value: 'test' },
          { type: 'number', value: '300' },
          { type: 'datetime', value: '2017-02-22T09:19:33.885Z' },
          { type: 'file', 
            value: { ref: 'file32', mimeType: 'text/plain', downloadUrl: 'download' } }
        ];
        expect(util.deflateValue(value, 'list')).toEqual([
          'test', 300, new Date('2017-02-22T09:19:33.885Z'),
          { ref: 'file32', mimeType: 'text/plain', downloadUrl: 'download' }
        ]);
      });
    });
    describe('when value is complex nested object', () => {
      it('should recursively convert value to plain object without field metadata', () => {
        const value = {
          foo: util.makeField({
            bar: util.makeField({
              baz: util.makeField('abc', 'text'),
              baq: util.makeField('20', 'number'),
              buz: util.makeField(false, 'boolean'),
              bez: util.makeField('2018-02-22T09:19:33.885Z', 'datetime'),
              bat: util.makeField(
                { ref: 'f2', mimeType: 'image/png', downloadUrl: 'down1' },
                'file')
            }, 'object'),
          },'object'),
          biz: util.makeField([
            { type: 'boolean', value: true },
            { type: 'text', value: 'xyz' },
            { type: 'datetime', value: '2018-02-24T09:20:33.885Z' },
            { type: 'number', value: '342.234' },
            { type: 'file', value: {
              ref: 'f3',
              mimeType: 'image/jpeg',
              downloadUrl: 'down3'
            } },
            { type: 'object', value: {
              fizz: util.makeField('buzz', 'text')
            } }
          ], 'list')
        };
        expect(util.deflateValue(value, 'object')).toMatchSnapshot();
      });
    });
  });

  describe('walkFieldPath', () => {
    let objField: Field;

    beforeEach(() => {
      objField = util.makeField({
        foo: util.makeField({
          fizz: util.makeField('buzz', 'text')
        }, 'object'),
        bar: util.makeField('bar value', 'text'),
        rabs: util.makeField([
          {
            type: 'boolean',
            value: true
          },
          {
            type: 'object',
            value: {
              dar: util.makeField('dar value', 'text')
            }
          }
        ], 'list')
      }, 'object');
    });

    describe('when the start field has type object', () => {
      it('should return nested field (fizz) at the specified path', () => {
        const res = util.walkFieldPath(objField, ['foo', 'fizz']);
        expect(res).toMatchSnapshot();
      });
      it('should walk through array indices and return dar field', () => {
        const res = util.walkFieldPath(objField, ['rabs', '1', 'dar']);
        expect(res).toMatchSnapshot();
      });
      it('should support array indices as ints (returns dar field)', () => {
        const res = util.walkFieldPath(objField, ['rabs', 1, 'dar']);
        expect(res).toMatchSnapshot();
      });
    });
    describe('when the start field is a list', () => {
      it('should return nested list item (returns boolean item)', () => {
        const value = objField.value as ObjectValue;
        const res = util.walkFieldPath(value.rabs, [0]);
        expect(res).toMatchSnapshot();
      });
      it('should return walk nested object and return dar field', () => {
        const value = objField.value as ObjectValue;
        const res = util.walkFieldPath(value.rabs, [1, 'dar']);
        expect(res).toMatchSnapshot();
      });
    });
    describe('when path is empty', () => {
      it('should return the input field', () => {
        const res = util.walkFieldPath(objField, []);
        expect(res).toEqual(objField);
      });
    });
    describe('when the path cannot be matched', () => {
      it('should return undefined', () => {
        expect(util.walkFieldPath(objField, ['foo', 'fizz', 'unknown'])).toBeUndefined();
        expect(util.walkFieldPath(objField, ['foo', 'unknown', 'fizz'])).toBeUndefined();
        expect(util.walkFieldPath(objField, ['random', 'path'])).toBeUndefined();
      });
      it('should return undefined when using array index out of bound', () => {
        expect(util.walkFieldPath(objField, ['rabs', 2, 'dar'])).toBeUndefined();
      });
    });
  });

  describe('walkEntityPath', () => {
    let entity: ApiEntity;
    beforeEach(() => {
      entity = {
        _id: 'e1',
        createdAt: '2018-02-22T09:19:33.885Z',
        updatedAt: '2018-02-22T09:19:33.885Z',
        createdBy: {
          _id: 'u1',
          type: 'user'
        },
        updatedBy: {
          _id: 'u2',
          type: 'user'
        },
        tags: [],
        data: {
          foo: util.makeField({
            fizz: util.makeField('buzz', 'text')
          }, 'object'),
          bar: util.makeField('bar value', 'text'),
          rabs: util.makeField([
            {
              type: 'boolean',
              value: true
            },
            {
              type: 'object',
              value: {
                dar: util.makeField('dar value', 'text')
              }
            }
          ], 'list')
        }
      };
    });
    it('should call walkFieldPath with the first field (returns dar field)', () => {
      const res = util.walkEntityPath(entity, ['rabs', 1, 'dar']);
      expect(res).toMatchSnapshot();
    });
    describe('when key in the path is not in entity', () => {
      it('should return undefined', () => {
        expect(util.walkEntityPath(entity, ['dabs', 1, 'dar'])).toBeUndefined();
        expect(util.walkEntityPath(entity, ['rabs', 2, 'dar'])).toBeUndefined();
        expect(util.walkEntityPath(entity, ['foo', 'barred'])).toBeUndefined();
      });
    });
    describe('when path is empty', () => {
      it('should return undefined', () => {
        expect(util.walkEntityPath(entity, [])).toBeUndefined();
      });
    });
  });

  describe('Expand query', () => {
    describe('Regex', () => {
      let query;
      beforeAll(async () => {
        query = {
          query: {
            name: 'Me myself',
            title: /developer/i
          },
          limit: 100,
          skip: 2
        };
      });
      it('should expand regex successfully', async () => {
        expect(util.expandQuery(query)).toMatchSnapshot();
      });
    });
    describe('Dates', () => {
      let query;
      beforeAll(async () => {
        query = {
          query: {
            name: 'Me myself',
            $or: [
              { time: { $gt: new Date() }},
              { age: { $lt: new Date() }}
            ]
          },
          limit: 100,
          skip: 2
        };
      });
      it('should expand regex successfully', async () => {
        expect(util.expandQuery(query)).toMatchSnapshot();
      });
    });
  });
});