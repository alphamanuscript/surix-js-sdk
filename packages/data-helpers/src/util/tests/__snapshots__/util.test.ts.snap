// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`util Expand query Dates should expand regex successfully 1`] = `
Object {
  "limit": 100,
  "query": Object {
    "$or": Array [
      Object {
        "time": Object {
          "$gt": Object {
            "$treatAsDatetime": "2019-06-18T20:55:26.711Z",
          },
        },
      },
      Object {
        "age": Object {
          "$lt": Object {
            "$treatAsDatetime": "2019-06-18T20:55:26.711Z",
          },
        },
      },
    ],
    "name": "Me myself",
  },
  "skip": 2,
}
`;

exports[`util Expand query Regex should expand regex successfully 1`] = `
Object {
  "limit": 100,
  "query": Object {
    "name": "Me myself",
    "title": Object {
      "$options": "i",
      "$regex": "/developer/i",
    },
  },
  "skip": 2,
}
`;

exports[`util deflateValue when value is complex nested object should recursively convert value to plain object without field metadata 1`] = `
Object {
  "biz": Array [
    true,
    "xyz",
    2018-02-24T09:20:33.885Z,
    342.234,
    Object {
      "downloadUrl": "down3",
      "mimeType": "image/jpeg",
      "ref": "f3",
    },
    Object {
      "fizz": "buzz",
    },
  ],
  "foo": Object {
    "bar": Object {
      "baq": 20,
      "bat": Object {
        "downloadUrl": "down1",
        "mimeType": "image/png",
        "ref": "f2",
      },
      "baz": "abc",
      "bez": 2018-02-22T09:19:33.885Z,
      "buz": false,
    },
  },
}
`;

exports[`util walkEntityPath should call walkFieldPath with the first field (returns dar field) 1`] = `
Object {
  "type": "text",
  "value": "dar value",
}
`;

exports[`util walkFieldPath when the start field has type object should return nested field (fizz) at the specified path 1`] = `
Object {
  "type": "text",
  "value": "buzz",
}
`;

exports[`util walkFieldPath when the start field has type object should support array indices as ints (returns dar field) 1`] = `
Object {
  "type": "text",
  "value": "dar value",
}
`;

exports[`util walkFieldPath when the start field has type object should walk through array indices and return dar field 1`] = `
Object {
  "type": "text",
  "value": "dar value",
}
`;

exports[`util walkFieldPath when the start field is a list should return nested list item (returns boolean item) 1`] = `
Object {
  "type": "boolean",
  "value": true,
}
`;

exports[`util walkFieldPath when the start field is a list should return walk nested object and return dar field 1`] = `
Object {
  "type": "text",
  "value": "dar value",
}
`;
